#
message("This is the FastqTrim() R Server Script")
message("Starting up")
# Index Input
option_list = list(
  optparse::make_option(c("--Forward_Fastq"), type="character", default="No Default",
                        help="The single-read or forward paired-end fastq file name (including .fastq file extension)",metavar = "character"),
  optparse::make_option(c("--Reverse_Fastq"), type="character", default="No Default",
                        help="When Paired_End_Data=TRUE, the reverse paired-end fastq file name (including .fastq file extension). This argument is ignored when Paired_End_Data=FALSE.",metavar = "character"),
  optparse::make_option(c("--Paired_End_Data"), type="logical", default=FALSE,
                        help="The input data type, where FALSE corresponds with single-end reads. [default=TRUE]",metavar = "logical"),
  optparse::make_option(c("--Adapter_Removal"), type="logical", default=FALSE,
                        help="Whether to remove adapters sequences defined with Adapters='example.txt' or 'sequence'. [default=FALSE]",metavar = "logical"),
  optparse::make_option(c("--Adapters"), type="character", default="No Default",
                        help="Either an R character string object of forward and reverse adapters or the .txt file of the adapters generated by FastqQual()",metavar = "character"),
  optparse::make_option(c("--Output_Base_Name"), type="character", default="Trimmed_Fastq",
                        help="The base name of the generated trimmed fastq files. No file extension is required. [default='Trimmed_Fastq']",metavar = "character"),
  optparse::make_option(c("--minLen"), type="integer", default=10,
                        help="(Optional) The minimum read length allowed. [default=20]",metavar = "integer"),
  optparse::make_option(c("--maxLen"), type="double", default=Inf,
                        help="The maximum read length allowed. [default=Inf, no limit]",metavar = "double"),
  optparse::make_option(c("--maxN"), type="integer", default=30,
                        help="maxN (Optional) The maximum number of unknown nucleotides N allowed within a single read. [default=30]",metavar = "integer"),
  optparse::make_option(c("--minQ"), type="integer", default=10,
                        help="(Optional) The minimum base quality score allowed. Reads with one or more bases below this threshold will be discarded. [default=10]",metavar = "integer"),
  optparse::make_option(c("--rm.phix"), type="logical", default=TRUE,
                        help="rm.phix (Optional) Whether to remove common contraminator sequences recorded in the phiX genome [default=TRUE]",metavar = "logical"),
  optparse::make_option(c("--truncQ"), type="integer", default=2,
                        help="(Optional) The minimum quality score of a base after which a read is truncated [default=2]",metavar = "integer"),
  optparse::make_option(c("--trimLeft"), type="integer", default=0,
                        help="(Optional) The number of nucleotides to remove on the 3' end of all reads. Use this if the first sequence cycles are of low quality. [default=0]",metavar = "integer"),
  optparse::make_option(c("--trimRight"), type="integer", default=0,
                        help="(Optional) The number of nucleotides to remove on the 5' end of all reads. Use this if the last sequence cycles are of low quality. [default=0]",metavar = "integer"),
  optparse::make_option(c("--maxEE"), type="double", default=Inf,
                        help="(Optional) The maximum expected error allowed for a single read. EE=sum(10^(-Q/10)). [default=Inf, No maximum]",metavar = "double"),
  optparse::make_option(c("--rm.lowcomplex"), type="integer", default=0,
                        help="(Optional) The minimum read complexity allowed, based on Shannon Entropy. [default=0]",metavar="integer")
)
opt_parser = optparse::OptionParser(option_list=option_list)
opt= optparse::parse_args(opt_parser)

# Build function
FastqTrim<-function(Forward_Fastq=opt$Forward_Fastq, Reverse_Fastq=opt$Reverse_Fastq, Paired_End_Data=opt$Paired_End_Data, Adapter_Removal=opt$Adapter_Removal, Adapters=opt$Adapters, Output_Base_Name=opt$Output_Base_Name, minLen=opt$minLen, maxLen=opt$maxLen, maxN=opt$maxN, minQ=opt$minQ, rm.phix=opt$rm.phix, truncQ=opt$truncQ, trimLeft=opt$trimLeft, trimRight=opt$trimRight, maxEE=opt$maxEE, rm.lowcomplex=opt$rm.lowcomplex){
  # Check Input
  if(file.exists(Forward_Fastq)==FALSE){return(message("No existing Fastq file has been submitted"))
  }else if(file.exists(Reverse_Fastq)==FALSE & Paired_End_Data==TRUE){return(message("No existing reverse fastq file has been submitted, but paired_end_data=TRUE"))
  }else if(Adapter_Removal==TRUE & Adapters=="No Default"){return(message("Adapter_Removal is set to true, but no adapter sequences have been provided. If adapters are unknown, perform a search with FastqQual first."))
  }

if(Paired_End_Data==FALSE){
  ## For Single Reads

  if(Adapter_Removal==TRUE){
    if(grepl(".txt",Adapters)==1 & .Platform$OS.type!="windows"){
      Adapters<-as.character(read.table(Adapters)[,1])
      Temp_Filtered_File_Name<-paste(Output_Base_Name,"_temp.fastq",sep="")

        message("Removing adapters specified in .txt file")
      Rbowtie2::remove_adapters(file1=Forward_Fastq,
                                adapter1 = Adapters[1],
                                output1=file.path(getwd(),Temp_Filtered_File_Name),
                                overwrite=TRUE,
                                "--threads 2")
    }else if(grepl(".txt",Adapters)!=1 & .Platform$OS.type!="windows"){
      Temp_Filtered_File_Name<-paste(Output_Base_Name,"_temp.fastq",sep="")

        message("Removing adapters specified in R character strings")
      Rbowtie2::remove_adapters(file1=Forward_Fastq,
                                adapter1 = Adapters[1],
                                output1=file.path(getwd(),Temp_Filtered_File_Name),
                                overwrite=TRUE,
                                "--threads 2")
    }else if(grepl(".txt",Adapters)==1 & .Platform$OS.type=="windows"){
      Adapters<-as.character(read.table(Adapters)[,1])
      Temp_Filtered_File_Name<-paste(Output_Base_Name,"_temp.fastq",sep="")

        message("Removing adapters specified in R character strings")
      command<-paste(sep="","'",.libPaths()[1],"/Rbowtie2/AdapterRemoval' --file1 ",Forward_Fastq," --adapter1 ",Adapters[1]," --output1 ",file.path(getwd(),Temp_Filtered_File_Name))
      system(command, intern=TRUE)
    }
    # Filter and Trim after adapter removal
      message("Filtering FastQ Files. This may take a while")
    Filtered_File_Name<-paste(Output_Base_Name,".fastq",sep="")
    dada2::filterAndTrim(fwd=Temp_Filtered_File_Name, filt=Filtered_File_Name,
                         truncQ=truncQ,
                         maxLen=maxLen,
                         minLen=minLen,
                         maxN=maxN,
                         minQ=minQ,
                         rm.phix=rm.phix,
                         truncLen=truncLen,
                         trimLeft=trimLeft,
                         trimRight=trimRight,
                         maxEE=maxEE,
                         rm.lowcomplex=rm.lowcomplex,
                         compress=FALSE, verbose=TRUE)
    message(" ")
    message("The Filtered FASTQ File has been saved in ", getwd()," as ", Filtered_File_Name)
    file.remove(Temp_Filtered_File_Name)

  }else{

  # Filter and Trim without adapter removal
  message("Filtering FastQ Files. This may take a while")
  Filtered_File_Name<-paste(Output_Base_Name,".fastq",sep="")
  dada2::filterAndTrim(fwd=Forward_Fastq, filt=Filtered_File_Name,
                       truncQ=truncQ,
                       maxLen=maxLen,
                       minLen=minLen,
                       maxN=maxN,
                       minQ=minQ,
                       rm.phix=rm.phix,
                       truncLen=truncLen,
                       trimLeft=trimLeft,
                       trimRight=trimRight,
                       maxEE=maxEE,
                       rm.lowcomplex=rm.lowcomplex,
                       compress=FALSE, verbose=TRUE)
  message(" ")
  message("The Filtered FASTQ File has been saved in ", getwd()," as ", Filtered_File_Name)
  }
} else{
  ## Paired End reads
  Forward_Filtered_File_Name<-paste(Output_Base_Name,"_fw.fastq",sep="")
  Reverse_Filtered_File_Name<-paste(Output_Base_Name,"_rv.fastq",sep="")

  # Remove Adapters
  if(Adapter_Removal==TRUE){
    if(grepl(".txt",Adapters)==1 & .Platform$OS.type!="windows"){
  Adapters<-as.character(read.table(Adapters)[,1])
  Temp_FW_Filtered_File_Name<-paste(Output_Base_Name,"temp_fw.fastq",sep="")
  Temp_RV_Filtered_File_Name<-paste(Output_Base_Name,"temp_rv.fastq",sep="")
    message("Removing adapters specified in .txt file")
  Rbowtie2::remove_adapters(file1=Forward_Fastq,
                            file2=Reverse_Fastq,
                            adapter1 = Adapters[1],
                            adapter2 = Adapters[2],
                            output1=file.path(getwd(),Temp_FW_Filtered_File_Name),
                            output2=file.path(getwd(),Temp_RV_Filtered_File_Name),
                            overwrite=TRUE,
                            "--threads 2")
    }else if(grepl(".txt",Adapters)!=1 & .Platform$OS.type!="windows"){
      Temp_FW_Filtered_File_Name<-paste(Output_Base_Name,"temp_fw.fastq",sep="")
      Temp_RV_Filtered_File_Name<-paste(Output_Base_Name,"temp_rv.fastq",sep="")
        message("Removing adapters specified in R character strings")
      Rbowtie2::remove_adapters(file1=Forward_Fastq,
                                file2=Reverse_Fastq,
                                adapter1 = Adapters[1],
                                adapter2 = Adapters[2],
                                output1=file.path(getwd(),Temp_FW_Filtered_File_Name),
                                output2=file.path(getwd(),Temp_RV_Filtered_File_Name),
                                overwrite=TRUE,
                                "--threads 2")
    }else if(grepl(".txt",Adapters)==1 & .Platform$OS.type=="windows"){
      Adapters<-as.character(read.table(Adapters)[,1])
      Temp_FW_Filtered_File_Name<-paste(Output_Base_Name,"temp_fw.fastq",sep="")
      Temp_RV_Filtered_File_Name<-paste(Output_Base_Name,"temp_rv.fastq",sep="")
        message("Removing adapters specified in R character strings")

      command<-paste(sep="","'",.libPaths()[1],"/Rbowtie2/AdapterRemoval' --file1 ",Forward_Fastq," --file2 ",Reverse_Fastq," --adapter1 ",Adapters[1]," --adapter2 ",Adapters[2]," --output1 ",file.path(getwd(),Temp_FW_Filtered_File_Name)," --output2 ",file.path(getwd(),Temp_RV_Filtered_File_Name))
      system(command, intern=TRUE)

    }else{
      Temp_FW_Filtered_File_Name<-paste(Output_Base_Name,"temp_fw.fastq",sep="")
      Temp_RV_Filtered_File_Name<-paste(Output_Base_Name,"temp_rv.fastq",sep="")
        message("Removing adapters specified in R character strings")

      command<-paste(sep="","'",.libPaths()[1],"/Rbowtie2/AdapterRemoval' --file1 ",Forward_Fastq," --file2 ",Reverse_Fastq," --adapter1 ",Adapters[1]," --adapter2 ",Adapters[2]," --output1 ",file.path(getwd(),Temp_FW_Filtered_File_Name)," --output2 ",file.path(getwd(),Temp_RV_Filtered_File_Name))
      system(command, intern=TRUE)
    }
  # Filter and Trim after adapter removal
    message("Filtering FastQ Files. This may take a while")
  dada2::filterAndTrim(fwd=Temp_FW_Filtered_File_Name, filt=Forward_Filtered_File_Name,
                       rev=Temp_RV_Filtered_File_Name, filt.rev=Reverse_Filtered_File_Name,
                       truncQ=truncQ,
                       maxLen=maxLen,
                       minLen=minLen,
                       maxN=maxN,
                       minQ=minQ,
                       rm.phix=rm.phix,
                       truncLen=truncLen,
                       trimLeft=trimLeft,
                       trimRight=trimRight,
                       maxEE=maxEE,
                       rm.lowcomplex=rm.lowcomplex,
                       compress=FALSE, verbose=TRUE)
  file.remove(Temp_FW_Filtered_File_Name)
  file.remove(Temp_RV_Filtered_File_Name)
    message(" ")
    message("The Filtered FASTQ Files have been saved in ", getwd()," as",   Forward_Filtered_File_Name, " and ",Reverse_Filtered_File_Name)
}else{
# Filter and Trim without adapter removal
  message("Filtering FastQ Files. This may take a while")
dada2::filterAndTrim(fwd=Forward_Fastq, filt=Forward_Filtered_File_Name,
                     rev=Reverse_Fastq, filt.rev=Reverse_Filtered_File_Name,
                     truncQ=truncQ,
                     maxLen=maxLen,
                     minLen=minLen,
                     maxN=maxN,
                     minQ=minQ,
                     rm.phix=rm.phix,
                     truncLen=truncLen,
                     trimLeft=trimLeft,
                     trimRight=trimRight,
                     maxEE=maxEE,
                     rm.lowcomplex=rm.lowcomplex,
                     compress=FALSE, verbose=TRUE)
  message(" ")
  message("The Filtered FASTQ Files have been saved in ", getwd()," as", Forward_Filtered_File_Name, " and ",Reverse_Filtered_File_Name)
}}
}

# Execute trimming
FastqTrim()
